pkg/ingester/ingester_query.go

```diff
@@ -0,0 +1,130 @@
+package ingester
+
+import (
+	"context"
+
+	"github.com/grafana/loki/pkg/iter"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/logqlmodel/stats"
+	"github.com/grafana/loki/pkg/util/math"
+)
+
+type IngesterQuery struct {
+	Id       uint32
+	instance *instance
+	loopAck  chan struct{}
+}
+
+func NewIngesterQuery(id uint32, instance *instance) *IngesterQuery {
+	ret := &IngesterQuery{
+		Id:       id,
+		instance: instance,
+		loopAck:  make(chan struct{}, 30),
+	}
+
+	// max 20 in flight
+	for i := 0; i < 20; i++ {
+		ret.loopAck <- struct{}{}
+	}
+
+	return ret
+}
+
+func (q *IngesterQuery) End() {
+	q.instance.queryMtx.Lock()
+	delete(q.instance.queries, q.Id)
+	q.instance.queryMtx.Unlock()
+}
+
+func (q *IngesterQuery) BackpressureWait(ctx context.Context) {
+
+	select {
+	case <-ctx.Done():
+		// The context is over, stop processing results
+		return
+	case <-q.loopAck:
+		// Process the results received
+	}
+}
+
+func (q *IngesterQuery) ReleaseAck() {
+	q.loopAck <- struct{}{}
+}
+
+// QuerierQueryServer is the GRPC server stream we use to send batch of entries.
+type QuerierQueryServer interface {
+	Context() context.Context
+	Send(res *logproto.QueryResponse) error
+}
+
+func (q *IngesterQuery) SendBatches(ctx context.Context, i iter.EntryIterator, queryServer QuerierQueryServer, limit int32) error {
+	stats := stats.FromContext(ctx)
+
+	// send until the limit is reached.
+	for limit != 0 && !isDone(ctx) {
```

Comment:
can the limit be negative? if so, I think it should be limit <= 0 instead

Reply:
if it is negative it's an error. Theres no precedent of assertions in loki that I've found so alas it's a silent failure.